Add reduce_on_plateau LR scheduler to contrib directory.

vz415 opened the pull request:
Following pull request #505 and issue #221 with @mtthss suggestions on changing where the scheduler is located (contrib) and using GradientTransformationExtraArgs API. Let me know if there's anything else that needs to be done to merge this. Cheers.

fabianp commented:
Thanks for the contribution @vz415 ! There seem to be some minor issues regarding formatting that break the tests (see https://github.com/google-deepmind/optax/actions/runs/6830352294).

Other than that this looks good to me. Green light to merge once the tests pass @mtthss ?

vz415 commented:
@fabianp fixed the spacing formatting issue 🤦‍♂️ and everything should be ready to merge.

vroulet commented on optax/contrib/prodigy.py:
```diff
@@ -46,7 +46,7 @@ class ProdigyState(NamedTuple):
     learning_rate: base.ScalarOrSchedule = 0.1,
     betas: tuple[float, float] = (0.9, 0.999),
-    beta3: float | None = None,
+    beta3: float = None,
```
Thanks for catching that! I fixed it, you'll need to merge with main.

fabianp commented:
Thanks @vz415 for the changes! A couple more minor things and then I think we're ready to merge:

1. Is there any paper we could cite that describes this approach? If so, please add them under "References" on the docstring (see examples on _src/alias.py)
2. Please add this new method to the API reference (docs/api.rst)

vz415 commented:
Hi @fabianp , I've addressed the issues below.
1. There's no seminal paper on the technique but here's the pytorch [documentation](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html). I can add some papers I've seen use it if wanted.
2. I added it to the API reference. Let me know if I need to edit it.

fabianp commented on docs/api.rst:
```diff
@@ -619,6 +619,7 @@ Schedules
 .. autofunction:: piecewise_constant_schedule
 .. autofunction:: piecewise_interpolate_schedule
 .. autofunction:: polynomial_schedule
+.. autofunction:: optax.contrib.reduce_on_plateau.reduce_on_plateau
```
```suggestion
.. autofunction:: optax.contrib.reduce_on_plateau
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,135 @@
+  min_improvement:float,
+  cooldown:int
+) -> base.GradientTransformationExtraArgs:
+  """ Args:
```
```suggestion
  """Reduce learning rate when a metric has stopped improving. 
  
  Models often benefit from reducing the learning once learning stagnates.
  This scheduler reads a metrics quantity and if no improvement is seen for
  a ‘patience’ number of epochs, the learning rate is reduced.
  
  Args:
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,135 @@
+def reduce_on_plateau(
+  reduce_factor: float,
+  patience: int,
+  min_improvement:float,
```
```suggestion
  min_improvement: float,
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,135 @@
+  reduce_factor: float,
+  patience: int,
+  min_improvement:float,
+  cooldown:int
```
```suggestion
  cooldown: int
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,135 @@
+  """ Args:
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning 
```
Why are these parameters in terms of number of epochs? Wouldn't it be more accurate to talk about number of iterations? It seems to me that the method has now knowledge of epochs, but keeps track of iterations instead

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,142 @@
+
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning 
```
```suggestion
  patience: Number of iterations with no improvement after which learning 
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,142 @@
+      rate will be reduced.
+  min_improvement: Threshold for measuring the new optimum, to only focus on 
+      significant changes.
+  cooldown: Number of epochs to wait before resuming normal operation 
```
```suggestion
  cooldown: Number of iterations to wait before resuming normal operation 
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,142 @@
+
+  Models often benefit from reducing the learning once learning stagnates.
+  his scheduler reads a metrics quantity and if no improvement is seen for
+  a ‘patience’ number of epochs, the learning rate is reduced
```
```suggestion
  a ‘patience’ number of epochs, the learning rate is reduced.
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,142 @@
+    current_loss = extra_args.get("loss")
+
+    # Check if the current loss is the best so far
+
```
```suggestion
```

fabianp commented on optax/contrib/reduce_on_plateau.py:
```diff
@@ -0,0 +1,142 @@
+  plateau_count: int
+  lr: float
+  cooldown_counter: int
+  cooldown:int
```
```suggestion
  cooldown: int
```

fabianp commented:
Is there a reason why the parameters reduce_factor, patience, etc. need to be passed to the state ReduceLROnPlateauState ?

fabianp commented:
Also, in its current state, I think this method is not straightforward to use (at least it was not clear to me). Would you be willing add a docstring or create an example on how to combine this scaling with (say) SGD or Adam?

vz415 commented:
Thanks for the review and feedback! Answering your question: those parameters are passed to update the learning rate after every epoch. Those parameters could externalized via functools.partial but I think this format makes it easier to implement for users, although more implicit in how the updates are carried out. Let me know if this makes sense or if you have other concerns.

I'll add a docstring with an example on how to use. If that's not clear I can create an example.

fabianp commented:
This now merged with some changes. Among other things, I changed some keyword arguments to match those of the pytorch implementation.

Other changes were done to fix some internal failures due to the fact that the internal pytype checks are a bit more stringent than the ones that run on the github actions.

If you feel that some of these changes are for the worst, please open another PR and propose modifications.

Regarding documentation, I mentioned earlier I think it is important to make an example on using this function. I created issue #679 to track progress on this

fabianp commented:
thanks for the answer @vz415 (and for your contribution)! I saw your comment after I had submitted. In any case please don't hesitate to open another PR to edit the submitted code
